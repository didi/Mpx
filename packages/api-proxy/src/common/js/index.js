export * from './web'
export * from './utils'
export * from './toPromise'