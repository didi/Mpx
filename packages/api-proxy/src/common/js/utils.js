/**
 *
 * @param {Object} options 原参数
 * @param {Object} updateOrRemoveOpt 要修改或者删除的参数
 * @param {Object} extraOpt 额外增加的参数
 * @returns {Object} 返回参数
 * @example
 * changeOpts({ a: 1, b: 2 }, {
 *  a: 'c', // a 变为 c
 *  b: '' // 删除 b
 * }, {
 *  d: 4 // 增加 d
 * })
 */
const hasOwnProperty = Object.prototype.hasOwnProperty

function type (n) {
  return Object.prototype.toString.call(n).slice(8, -1)
}

function hasOwn (obj, key) {
  return hasOwnProperty.call(obj, key)
}

function changeOpts (options, updateOrRemoveOpt = {}, extraOpt = {}) {
  let opts = {}

  Object.keys(options).forEach(key => {
    const myKey = hasOwn(updateOrRemoveOpt, key) ? updateOrRemoveOpt[key] : key
    if (myKey !== '') {
      opts[myKey] = options[key]
    }
  })

  opts = Object.assign({}, opts, extraOpt)
  return opts
}

/**
 * @param {Object} opts 原参数
 * @param {Function} getOptions 获取 success 回调修改后的参数
 * @param {Object} thisObj this对象
 */
const handleSuccess = (opts, getOptions = noop, thisObj) => {
  if (!opts.success) {
    return
  }
  const _this = thisObj || this
  const cacheSuc = opts.success
  opts.success = res => {
    const changedRes = getOptions(res) || res
    cacheSuc.call(_this, changedRes)
  }
}

function getEnvObj () {
  switch (__mpx_mode__) {
    case 'wx':
      return wx
    case 'ali':
      return my
    case 'swan':
      return swan
    case 'qq':
      return qq
    case 'tt':
      return tt
    case 'jd':
      return jd
    case 'qa':
      return qa
    case 'dd':
      return dd
    case 'web':
    case 'ios':
    case 'android':
      return {}
  }
}

function warn (msg) {
  console.warn && console.warn(`[@mpxjs/api-proxy warn]:\n ${msg}`)
}

function error (msg) {
  console.error && console.error(`[@mpxjs/api-proxy error]:\n ${msg}`)
}
function envError (method) {
  return () => {
    console.error && console.error(`[@mpxjs/api-proxy error]:\n ${__mpx_mode__}环境不支持${method}方法`)
  }
}

function noop () {
}

function makeMap (arr) {
  return arr.reduce((obj, item) => {
    obj[item] = true
    return obj
  }, {})
}

function parseDataset (dataset) {
  const parsed = {}
  for (const key in dataset) {
    if (hasOwn(dataset, key)) {
      try {
        parsed[key] = JSON.parse(dataset[key])
      } catch (e) {
        parsed[key] = dataset[key]
      }
    }
  }
  return parsed
}

function defineUnsupportedProps (resObj, props) {
  const defineProps = {}
  props.forEach((item) => {
    defineProps[item] = {
      get () {
        warn(`The ${item} attribute is not supported in ${__mpx_mode__} environment`)
        return null
      }
    }
  })
  Object.defineProperties(resObj, defineProps)
}

const isBrowser = typeof window !== 'undefined'

function throwSSRWarning (info) {
  console.error(`[Mpx runtime error]: Dangerous API! ${info}, It may cause some problems, please use this method with caution`)
}

const ENV_OBJ = getEnvObj()

export {
  changeOpts,
  handleSuccess,
  getEnvObj,
  error,
  envError,
  warn,
  noop,
  makeMap,
  isBrowser,
  hasOwn,
  throwSSRWarning,
  ENV_OBJ,
  parseDataset,
  type,
  defineUnsupportedProps
}
