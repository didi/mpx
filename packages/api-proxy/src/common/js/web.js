function isTabBarPage (url, router) {
  const tabBarPagesMap = global.__tabBarPagesMap
  if (!tabBarPagesMap || !url) return false
  const path = router.match(url, router.history.current).path
  return !!tabBarPagesMap[path.slice(1)]
}

/**
 * Creates a new DOM element with the specified tag, attributes, and children.
 *
 * @param {string} tag - The tag name of the new element.
 * @param {Object.<string, string>} [attrs={}] - An object containing the attributes to set on the new element.
 * @param {Array.<HTMLElement>} [children=[]] - An array of child elements to append to the new element.
 * @returns {HTMLElement} The newly created DOM element.
 */
function createDom (tag, attrs = {}, children = []) {
  const dom = document.createElement(tag)
  Object.keys(attrs).forEach(k => dom.setAttribute(k, attrs[k]))
  children.length && children.forEach(child => dom.appendChild(typeof child === 'string' ? document.createTextNode(child) : child))
  return dom
}

// 在H5中，直接绑定 click 可能出现延时问题，很多点击可以关闭的组件被唤出之后，有概率立马触发点击事件，导致组件被关闭。
// 使用该方法通过 touchstart 和 touchend 模拟 click 事件，解决延时问题。
function bindTap (dom, handler) {
  dom.addEventListener('tap', handler)
  return () => {
    dom.removeEventListener('tap', handler)
  }
}

/**
 * 获取弹窗应当挂载的根节点
 * @returns dom
 */
function getRootElement () {
  const page = getCurrentPages().slice(-1)[0]?.$el
  return page || document.body
}

export {
  createDom,
  bindTap,
  getRootElement,
  isTabBarPage
}
