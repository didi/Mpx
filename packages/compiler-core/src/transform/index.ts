import { type ExpressionNode } from '../ast'

export interface ImportItem {
  exp: string | ExpressionNode
  path: string
}

// export interface TransformContext {
//   /** */
// }

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type TransformContext = any
