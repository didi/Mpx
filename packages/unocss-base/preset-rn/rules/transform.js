// transform-gpu ? preserve-3d ?
