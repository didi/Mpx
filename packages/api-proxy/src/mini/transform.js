import { error, getEnvObj, genFromMap, makeMap } from '../common/js'
import getWxToAliApi from './platform/wxToAli'

const fromMap = genFromMap()

function joinName (from = '', to = '') {
  return `${fromMap[from]}_${to}`
}

function transformApi (options) {
  const envObj = getEnvObj()
  const wxToAliApi = getWxToAliApi({ optimize: options.optimize })
  const platformMap = {
    'wx_ali': wxToAliApi,
    'qq_ali': wxToAliApi,
    'swan_ali': wxToAliApi,
    'tt_ali': wxToAliApi
  }
  const needProxy = {}
  const excludeMap = makeMap(options.exclude)
  // 后续不一定只转换ali，需要基于目标平台决定合并的自定义api集，例如输出快应用时需要定义并合并wxToQaApi
  Object.keys(envObj).concat(Object.keys(wxToAliApi)).forEach((key) => {
    if (!excludeMap[key]) {
      needProxy[key] = envObj[key] || wxToAliApi[key]
    }
  })
  const transedApi = Object.create(null)
  Object.keys(needProxy).forEach(api => {
    // 非函数不做转换
    if (typeof needProxy[api] !== 'function') {
      transedApi[api] = needProxy[api]
      return
    }

    transedApi[api] = (...args) => {
      const to = options.to
      let from = options.from

      if (args.length > 0) {
        from = args.pop()
        if (typeof from !== 'string' || !fromMap[from]) {
          args.push(from)
          from = options.from
        }
      }

      const fromTo = joinName(from, to)
      if (options.custom[fromTo] && options.custom[fromTo][api]) {
        return options.custom[fromTo][api].apply(this, args)
      } else if (
        platformMap[fromTo] &&
        platformMap[fromTo][api]
      ) {
        return platformMap[fromTo][api].apply(this, args)
      } else if (envObj[api]) {
        return envObj[api].apply(this, args)
      } else {
        error(`当前环境不存在 ${api} 方法`)
      }
    }
  })

  return transedApi
}

export default transformApi
