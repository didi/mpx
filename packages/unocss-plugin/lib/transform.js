import MagicString from 'magic-string'
import transformerDirectives from '@unocss/transformer-directives' //default
import { getReplaceSource } from './source.js'
const escapedReg = /\\(.)/g

function mpEscape (str, escapeMap = {}) {
  return str.replace(escapedReg, (_, p1) => {
    if (escapeMap[p1]) return escapeMap[p1]
    // unknown escaped
    return escapeMap.unknown
  })
}

function escapeKey (str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function buildAliasTransformer (alias) {
  if (!alias || !Object.keys(alias).length) {
    return s => getReplaceSource(s)
  }

  const keys = Object.keys(alias).sort((a, b) => b.length - a.length).map(i => escapeKey(i)).join('|')
  const regexText = `\\*(?:${keys})(?<=[^w-])`
  const regex = new RegExp(regexText, 'g')
  return function transformAlias (source) {
    source = getReplaceSource(source)
    const content = source.original().source()
    let match
    regex.lastIndex = 0
    while (match = regex.exec(content)) {
      const start = match.index
      const end = start + match[0].length - 1
      const name = content.slice(start + 1, end + 1)
      const replacement = alias[name]
      source.replace(start, end, replacement)
    }
    return source
  }
}
const regexCache = {}
function makeRegexClassGroup (separators = ['-', ':']) {
  const key = separators.join('|')
  if (!regexCache[key]) regexCache[key] = new RegExp(`((?:[!@<~\\w+:_/-]|\\[&?>?:?\\S*\\])+?)(${key})\\(((?:[~!<>\\w\\s:/\\\\,%#.$?-]|\\[.*?\\])+?)\\)(?!\\s*?=>)`, 'gm')
  regexCache[key].lastIndex = 0
  return regexCache[key]
}

function transformGroups (source, options = {}) {
  source = getReplaceSource(source)
  const content = source.original().source()
  let match
  const groupReg = makeRegexClassGroup(options.separators)
  while (match = groupReg.exec(content)) {
    const start = match.index
    const [text, a, spread, b] = match
    const end = start + text.length - 1
    const replacement = b.split(/\s+/g).filter(Boolean).map(i => i.replace(/^(!?)(.*)/, `$1${a}${spread}$2`)).join(' ')
    source.replace(start, end, replacement)
  }
  return source
}

const hasDirectiveTest = /@(apply|screen|layer)\s/
const hasThemeFunctionTest = /theme\(.*?\)/

function cssRequiresTransform (source) {
  return hasDirectiveTest.test(source) || hasThemeFunctionTest.test(source)
}

async function transformStyle (
  code,
  id,
  uno
) {
  const s = new MagicString(code)
  await transformerDirectives().transform(s, id, { uno })
  if (s.hasChanged()) {
    code = s.toString()
  }
  return code
}

export {
  cssRequiresTransform,
  transformGroups,
  mpEscape,
  transformStyle,
  buildAliasTransformer
}
