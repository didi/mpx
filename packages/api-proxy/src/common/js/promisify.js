import { ENV_OBJ } from './utils'

// 特别指定的不进行Promise封装的方法
const blackList = [
  'clearStorage',
  'hideToast',
  'hideLoading',
  'drawCanvas',
  'canIUse',
  'stopRecord',
  'pauseVoice',
  'stopVoice',
  'pauseBackgroundAudio',
  'stopBackgroundAudio',
  'showNavigationBarLoading',
  'hideNavigationBarLoading',
  'createAnimation',
  'createAnimationVideo',
  'createSelectorQuery',
  'createIntersectionObserver',
  'getPerformance',
  'hideKeyboard',
  'stopPullDownRefresh',
  'createWorker',
  'pageScrollTo',
  'reportAnalytics',
  'getMenuButtonBoundingClientRect',
  'reportMonitor',
  'createOffscreenCanvas',
  'reportEvent',
  'connectSocket',
  'base64ToArrayBuffer',
  'getDeviceInfo',
  'getWindowInfo'
]

function getMapFromList (list) {
  if (list && list.length) {
    const map = {}
    list.forEach((item) => {
      map[item] = true
    })
    return map
  }
}

function promisify (listObj, whiteList, customBlackList) {
  const result = {}
  const whiteListMap = getMapFromList(whiteList)
  const blackListMap = getMapFromList(blackList.concat(customBlackList))

  function promisifyFilter (key) {
    if (whiteListMap && whiteListMap[key] !== undefined) {
      return !!whiteListMap[key]
    } else {
      return !(blackListMap[key] || // 特别指定的方法
        /^get\w*Manager$/.test(key) || // 获取manager的api
        /^create\w*Context$/.test(key) || // 创建上下文相关api
        /^(on|off)/.test(key) || // 以 on* 或 off开头的方法
        /\w+Sync$/.test(key) // 以Sync结尾的方法
      )
    }
  }

  Object.keys(listObj).forEach(key => {
    if (typeof listObj[key] !== 'function') {
      return
    }

    result[key] = function (...args) {
      const obj = args[0] || {}
      // 不需要转换 or 用户已定义回调，则不处理
      if (!promisifyFilter(key)) {
        return listObj[key].apply(ENV_OBJ, args)
      } else { // 其他情况进行转换
        if (!args[0]) args.unshift(obj)
        let returned
        const promise = new Promise((resolve, reject) => {
          const originSuccess = obj.success
          const originFail = obj.fail
          obj.success = function (res) {
            originSuccess && originSuccess.call(this, res)
            resolve(res)
          }
          obj.fail = function (e) {
            originFail && originFail.call(this, e)
            reject(e)
          }
          returned = listObj[key].apply(ENV_OBJ, args)
        })
        promise.__returned = returned
        return promise
      }
    }
  })

  return result
}

export default promisify
