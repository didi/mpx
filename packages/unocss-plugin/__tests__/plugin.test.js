import MpxUnocssPlugin from '../lib/index'
import { describe, expect, test } from 'vitest'
import { getRawSource } from '../lib/source.js'
import { e as cssEscape } from '@unocss/core'
import { mpEscape } from '../lib/transform.js'
import presetMpx from '@mpxjs/unocss-base/lib/index'
// const { presetLegacyCompat } = require('@unocss/preset-legacy-compat')

// import testpage from './123.mpx?resolve'
describe('test plugin', async () => {
  const mode = 'wx'
  const plugin = new MpxUnocssPlugin({
    config: {
      presets: [
        presetMpx()
      ]
    }
  })
  const mockCompilation = {
    fileDependencies: new Set()
  }
  const uno = await plugin.createContext(mockCompilation, mode)
  const parseTemplate = plugin.getTemplateParser(uno)
  async function testTemplate (content, generateOptions = {
    preflights: false,
    safelist: false
  }) {
    const source = getRawSource(content)
    const classmap = {}
    const { newsource } = parseTemplate(source, (className) => {
      if (!className) {
        return className
      }
      classmap[className] = true
      return mpEscape(cssEscape(className), plugin.options.escapeMap)
    })
    // 测试模板是否转义
    expect(newsource.source()).toMatchSnapshot()
    const classList = Object.keys(classmap)
    // 测试类名是否正确识别
    expect(classList).toMatchSnapshot()
    const list = new Set(classList)
    const unoFileContent = await plugin.generateStyle(uno, classList, generateOptions)
    // 测试css输出是否正确
    expect(unoFileContent).toMatchSnapshot()
  }
  test('test-template', async () => {
    await testTemplate('<view class="translate-[-50%,-50%] text-12px bg-#fff/10" />')
    await testTemplate('<view wx:class="{{[\'translate-[-50%]\',{\'text-12px text-16px\': false},\'bg-#fff/10\']}}" />')
  })
})
