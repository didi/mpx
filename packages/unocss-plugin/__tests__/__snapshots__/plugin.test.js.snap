// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test plugin > test-template 1`] = `"<view class="translate-_bl_-50_p__2c_-50_p__br_ text-12px bg-_h_fff_s_10" />"`;

exports[`test plugin > test-template 2`] = `
[
  "translate-[-50%,-50%]",
  "text-12px",
  "bg-#fff/10",
]
`;

exports[`test plugin > test-template 3`] = `
"/* layer: default */
.translate-_bl_-50_p__2c_-50_p__br_{--un-translate-x:-50%;--un-translate-y:-50%;transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
.bg-_h_fff_s_10{background-color:rgb(255 255 255 / 0.1);}
.text-12px{font-size:12px;}"
`;

exports[`test plugin > test-template 4`] = `"<view wx:class="{{(['translate-_bl_-50_p__br_',{'text-12px': false},'bg-_h_fff_s_10'])}}" />"`;

exports[`test plugin > test-template 5`] = `
[
  "translate-[-50%]",
  "text-12px",
  "bg-#fff/10",
]
`;

exports[`test plugin > test-template 6`] = `
"/* layer: default */
.translate-_bl_-50_p__br_{--un-translate-x:-50%;--un-translate-y:-50%;transform:translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z)) rotate(var(--un-rotate)) rotateX(var(--un-rotate-x)) rotateY(var(--un-rotate-y)) rotateZ(var(--un-rotate-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z));}
.bg-_h_fff_s_10{background-color:rgb(255 255 255 / 0.1);}
.text-12px{font-size:12px;}"
`;
